lib/rules/no-else-return.js

```diff
@@ -80,15 +80,30 @@ module.exports = function(context) {
     /**
      * Check the consequent/body node to make sure it is not
      * a ReturnStatement or an IfStatement that returns on both
-     * code paths.  If it is, display the context report.
+     * code paths.
      *
      * @param {Node} node The consequent or body node
      * @param {Node} alternate The alternate node
-     * @returns {void}
+     * @returns {boolean} `true` if it is a Return/If node that always returns.
+     */
+    function checkForReturnOrIf(node) {
+        return checkForReturn(node) || checkForIf(node);
+    }
+
+
+    /**
+     * Check whether a node returns in every codepath.
+     * @param {Node} node The node to be checked
+     * @returns {boolean} `true` if it returns on every codepath.
      */
-    function checkForReturnOrIf(node, alternate) {
-        if (checkForReturn(node) || checkForIf(node)) {
-            displayReport(alternate);
+    function alwaysReturns(node) {
+        // If we have a BlockStatement, check each consequent body node.
+        if (node.type === "BlockStatement") {
+            return node.body.length > 0 && node.body.some(checkForReturnOrIf);
```

Comment:
Good catch. I'm not sure what to do about the iteration. Notice that this implementation is only slightly more expensive than the previous one (because we now iterate over all `if-else-if-...` blocks). Also, `checkForIf` is already a "naive" implementation, that does not check exhaustively for `return` in every codepath.

I guess that ideally we should reuse the `estraverse` traversal to do the job for us. However, I can't see a way to do that without the ability to add metadata to nodes, or having a unique id per node (although a sort of uuid could be generated by the node range/location).